src/System.Reflection.MetadataLoadContext/src/System/Reflection/PathAssemblyResolver.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+
+namespace System.Reflection
+{
+    /// <summary>
+    /// A MetadataAssemblyResolver that uses paths to each assembly.
+    /// </summary>
+    /// <remarks>
+    /// If the corresponding MetadataLoadContext does not have the CoreAssemblyName property set,
+    /// default values are provided which are compatible with .NET Core and .NET Framework.
+    /// </remarks>
+    public class PathAssemblyResolver : MetadataAssemblyResolver
+    {
+        private Dictionary<string, string> _fileToPaths = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+        public PathAssemblyResolver(IEnumerable<string> assemblyPaths)
+        {
+            if (assemblyPaths == null)
+                throw new ArgumentNullException(nameof(assemblyPaths));
+
+            foreach (string path in assemblyPaths)
+            {
+                string file = Path.GetFileNameWithoutExtension(path);
+                _fileToPaths.Add(file, path);
```

Comment:
@nguerrera for mainstream scenarios, I don't see the same assembly by name being loaded more than once for inspection purposes, or care to distinguish strong names. .NET Core, for example, works well with the current approach as both take a simple list of paths to assemblies.

Do you have any (mainstream-possible) scenarios?

Options:
1) Do nothing; someone could add their own `AssemblyNameAndDirectoryResolver` that takes a list of AssemblyNames and a corresponding list of directories.
2) Add a new constructor such as:
```public PathAssemblyResolver(IEnumerable<string> assemblyNames, IEnumerable<string> directories)```
where each assembly name can be simple (just name) or strong (including version, publickeytoken, culture).

However, for incoming calls to Resolve() we'd have to add policy to determine whether to key off the exact version + publickeytoken + culture, or be flexible and accept >= version, simple name only, etc.

Reply:
Just want to chime in here, that a more sophisticated default resolver would be **greatly** useful for anyone doing anything but the simplest of assembly loading. Mono.Cecil's resolver is quite simple, and only handles a few default cases … so it's been quite a task to write one from scratch for my project. I would implore that @nguerrera's thoughts here be considered 😄  👍

Reply:
> I don't understand why the inputs to the resolver would have to change to support distinct versions / strong names.

They wouldn't have to change - just an example of a very explicit model from the caller without the resolver having to inspect each assembly...

Closer to what you were thinking would be to keep the same constructor, but use a list not a hashtable. The Resolve() method would then use any "knobs" to control what it needs to do (like version >=, pick closest culture, etc) and will have to load\inspect each assembly (with same name) to determine best fit.

In either case though, yes I will add your two minimum asks (documentation, better "duplicate assembly" exception).

Reply:
@joelmartinez are you referring to matching run-time binding semantics or having better inspection-only\reflection-only binding semantics?

Since the runtime assembly binding logic in CoreClr (binding redirects, default search locations) is quite differnet from NetFx (explicit paths determined by parsing deps.json files), yes it is not easy to write a general-purpose binder that matches runtime semantics. Here's one [request ](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcorefx%2Fissues%2F24026&data=02%7C01%7CSteve.Harter%40microsoft.com%7C2104373926584068aebf08d625722fa7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636737569717813299&sdata=OIafuvTmMIcwF2wIfJ4hLC1AKDX16cowIY8TRmnGkW0%3D&reserved=0)to expose the internal functionality.